contracts/starknet/src/compute_engines/summary_stats/library.cairo

```diff
@@ -41,24 +42,65 @@ namespace SummaryStats {
         return _mean;
     }
 
+    func calculate_skip_frequency{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        total_samples: felt, num_samples: felt
+    ) -> felt {
+        alloc_locals;
+        local skip_frequency;
+
+        let (q, r) = unsigned_div_rem(total_samples, num_samples);
+        if (q == 0) {
+            return 1;
+        }
+
+        let is_valid = is_le(r * 2, num_samples);
+        if (is_valid == 1) {
+            skip_frequency = q;
+        } else {
+            skip_frequency = q + 1;
+        }
+        return skip_frequency;
+    }
+
     func calculate_volatility{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
-        oracle_address: felt, key: felt, start_tick: felt, end_tick: felt
+        oracle_address: felt, key: felt, start_tick: felt, end_tick: felt, num_samples: felt
     ) -> felt {
         alloc_locals;
+        let is_valid_size = is_le(num_samples, 200);
+
+        with_attr error_message("num_samples is too large.  Must be <= 200") {
+            assert is_valid_size = TRUE;
+        }
+
         let (latest_checkpoint_index) = IOracle.get_latest_checkpoint_index(
             contract_address=oracle_address, key=key
         );
-        let (cp, start_index) = IOracle.get_last_spot_checkpoint_before(
+        let (_start_cp, start_index) = IOracle.get_last_spot_checkpoint_before(
             oracle_address, key, start_tick
         );
+        local end_index;
+        if (end_tick == 0) {
+            end_index = latest_checkpoint_index;
+        } else {
+            let (_end_cp, _end_idx) = IOracle.get_last_spot_checkpoint_before(
+                oracle_address, key, end_tick
+            );
+            end_index = _end_idx;
+        }
 
         with_attr error_message("Not enough data") {
             assert_not_equal(start_index, latest_checkpoint_index);
         }
+
         let (tick_arr: TickElem**) = alloc();
-        _make_array(0, oracle_address, key, latest_checkpoint_index, start_index, tick_arr);
+        // TODO: why is end index first?
```

Comment:
?

Reply:
> Very hard to check the math and understand what was broken last time and what was fixed this time, but looks good overall

wasn't broken, we updated the volatility to use `N-1` instead of `N` as a divisor for degrees of freedom.  update from sara. she can provide context if you are curious.